### PR TITLE
New Mini-Orchestration Module - hubPeeredSpoke

### PR DESCRIPTION
**Pull request by KiZach**
PR for this feature request https://github.com/Azure/ALZ-Bicep/issues/173

Move subscription to target management group, using subPlacement module
Create Resource Group in subscription, using resourceGroup module
Create VNET in subscription, using spokeNetworking module
Peer VNET to hub VNET, using vnetPeering module
Peer VNET to virtual wan hub, using vnetPeeringVwan module

### Breaking Changes

None

image

image

## As part of this Pull Request I have

- [X] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [X] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [x] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [X] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [X] Performed testing and provided evidence.
- [ ] Updated tests *(if required)* [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml) - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows) - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
- [X] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md:3]:
```suggestion
This module acts as an orchestration module that create and configures a spoke network to deliver the Azure Landing Zone Hub & Spoke architecture, for both traditional Hub & Spoke and Virtual WAN, which is also described in the wiki on the [Deployment Flow article](https://github.com/Azure/ALZ-Bicep/wiki/DeploymentFlow).
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md:1]:
```suggestion
# Module: Orchestration - hubPeeredSpoke - Spoke network, including peering to Hub (Hub & Spoke or Virtual WAN)
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:1]: Please remove the multi-line comment section at the top of this `.bicep` file.

We have recently moved away from this and have update the contribution guide: https://github.com/Azure/ALZ-Bicep/wiki/Contributing#bicep-file-structure-example

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:12]: Can we default this to the deployment location or the resource group location instead of a fixed location, like we do in other modules?

Please update the `README.md` also 👍

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:30]: Lets remove this list of '@allowed' inputs as whilst I like its intent, we should keep the module use open and flexible to as many users as possible 👍

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:53]:
```suggestion
param parDnsServerIpArray array = []
```
we are adopting camelCasing throughout all our modules in an upcoming PR so this just gets us a little ahead 👍

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:56]:
```suggestion
param parNextHopIpAddress string = ''
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:59]:
```suggestion
param parBgpRoutePropagation bool = false
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:161]:
```suggestion
    parBGPRoutePropagation: parBgpRoutePropagation
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:66]:
```suggestion
param parHubVirtualNetworkId string
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:36]:
```suggestion
param parPeeredVnetSubscriptionMgPlacement string = ''
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:72]:
```suggestion
param parAllowHubVpnGatewayTransit bool = false
```

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:79]: Think this will need amending as we are remove the "allowed" list for this parameter?

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:118]: Need to make this more unique so it is not the same for every spoke VNet.

Maybe add the spoke subscription ID into the unique string piece?

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md:13]: Can we add a note here that you can only choose the traditional Hub & Spoke method OR VWAN and not both?

Just to avoid confusion where someone tries to do traditional peering and also VWAN connection to a hub?

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Awesome work here @KiZach thanks so much for your efforts here, greatly appreciated 👍

Just a few changes required if you could make them and then we will review again 👍

Thanks

Jack

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:12]: Resource group creation is part of the module so I think we need this parameter right?

Comment (jtracey93) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:12]: Correct we need the parameter but it's default value should be changed as per previous comment. 👍

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in newest commit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md:13]: Updated in newest commit

**Comment by jtracey93** [comment]
Thanks @KiZach will review again shortly.

Can you confirm if the `targetScope` for this bicep file is correct? Currently it's not set so Resource Group scope, should it be set to Subscription scope?

**Comment by KiZach** [comment]
> Thanks @KiZach will review again shortly.
> 
> Can you confirm if the `targetScope` for this bicep file is correct? Currently it's not set so Resource Group scope, should it be set to Subscription scope?

@jtracey93 As the subscriptionPlacement module is used, the targetScope is managementGroup.

I will retest with the code comitted today as well.

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep:96]: Could you move this to the top of the file as per the Contribution Guide for consistency with other modules? 👍

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (jtracey93) [inline infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md:22]:
```suggestion
 | parLocation                            | string | Empty string `''`                                               | The region to deploy all resources into                                | Valid Azure Region | `deployment().location`                                                                                                                                              |
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Thanks @KiZach a couple more pieces and if you could re-test and post results in a comment on the PR as evidence that would be great 👍👍👍


Great work here 👏

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/hubPeeredSpoke.bicep]: Updated in ThirdCommit

File: infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md
Comment (KiZach) [reply infra-as-code/bicep/orchestration/hubPeeredSpoke/README.md]: Updated in ThirdCommit

**Comment by jtracey93** [comment]
> @jtracey93 Jack Tracey FTE fixed the 2 issues, but also found some breaking changes by the camelCasing committed today.
> 
> Proff of tests:
> 
> image
> 
> image

Thanks @KiZach, looking good. 

We have our internal scrum in an hour or so, so we will review as team and likely merge this evening 👍

**Comment by jtracey93** [comment]
Hey @KiZach,

Update as promised, we are cool with it, but we are going to get another set of eyes on this in the next few days just to ensure we are happy.

@JamJarchitect will get to this on Monday next week to give his review/approval and then we will merge.

#AllInTheNameOfQualityAssurance 👍

**Comment by jtracey93** [comment]
@msftbot require approvals from @jtracey93 and @JamJarchitect on this PR

**Review by JamJarchitect — APPROVED** [review]
LGTM. Great module and contribution! 👏

JJ

**Review by jtracey93 — APPROVED** [review]

**Pull request merged**
